### ISSUE TITLE
Force sources for add_library onto separate lines

### ISSUE THREAD

**Issue by yeganer** [cheshirekow/cmake_format#97]
Hey,
I just found this tool and it's amazing! However I have an issue that bothers me.

If the arguments to add_library fit onto one line, they'll get mashed together like so:
```
add_library(
    foo src/file1.c src/file2.c
)
```
However if the list of files is long enough the lines are wrapped like so:
```
add_library(
    foo
    src/file1.c
    src/file2.c
    src/file3.c
    src/file4.c
    src/file5.c
)
```
I want consistent formatting like the 2nd example, even if all arguments would fit in one line. That way adding a file to the list of sources does not cause a sudden wrap of all arguments which results in an ugly diff in git.

I've tried changing the algorithm order but that messes up the formatting for other commands. Same with changing `max_subargs_per_line`.
Any suggestions on how to achieve the desired formatting?
Looking at the code suggests my request is not possible right now but you are planning to expand the per-command settings.
In general I think it would be great if we could specify for each argument if it should be wrapped or not.

What I this means is configuring add_library to have `pargs = [1, '+']` and then somehow configure the `+` to always wrap. Maybe like that: `pargs = [1, {'value': '+', 'always_wrap': True}]`.
I can think of several other commands where this kind of fine-grained control would be useful. For example one could want to wrap the PUBLIC and PRIVATE libraries in `target_link_libraries`.

It would be really cool to have such a feature. If you think my proposal is a good solution, I could try to make a prototype.

**Comment by cheshirekow** [comment]
There is already some work in progress on almost exactly what you are suggesting. The plan is to make options like `algorithm_order` and `max_subpargs_per_line` settable on a per-command basis. In addition, each argument group (KWARG or PARG) is treated the same as a command, so you could change the settings for a particular argument group within a statement. 

One thing that is new in your suggestion is to change the `pargs` specification to a list, `[1, '+']`, enabling us to separate positional arguments into groups... potentially with different formatting options. I'll see if I can work this into the design. This idea is actually pretty powerful, and presents a solution to a problem I ran into attempting to implement #78 as well. So thanks for sharing it! I definitely think it's a good idea.

**Comment by yeganer** [comment]
Thanks for the response.
I just had another idea but I'm not sure how do-able that is. The specification for arguments currently includes numbers and '+'/'*' as wildcards to specify how many following arguments a kwarg/parg takes. One could add another type 'f' which would mark an ArgGroup as a FileGroup and wrapping.

I haven't looked at the source code in detail to get a feeling how difficult it would be to implement that.
In my mind this feature could solve some issues and would be easier to configure than per-command/per-argument based settings of `algorithm_order`.

What do you think?

**Comment by cheshirekow** [comment]
It's a simpler idea, and it seems like it could solve this particular problem... but it would require custom case logic in the formatting internals. 

I think per-command configuration has broader feature coverage, solving your problem as well as others.  There are two outstanding feature requests on github that it would enable,  and at least one issue I've been wanting to deal with in my own list files. It also has the advantage of not requiring additional formatting logic... which is good. 

Between the two solution options I think per-command formatting is the better one. 

That said, your idea of `f` is interesting. I can see an advantage in signifying that a group of arguments is a group of "filenames". Coming around a bit full circle, the implementation of `f` could be "An argument group with per-command algorithm order set to this other default thing", so that it's a kind of short-hand for a common override... which could make sense. 

Alternatively, instead of being a shorthand for an override, I could also add a "tags" field to the command specification. Tags can be used to alter the behavior of the different formatting algorithms. The problem I see with that is that it adds another "configuration axis" which might complicated interactions down the road. I'm inclined to avoid something like "tags" for now.

**Comment by cheshirekow** [comment]
P.S. it occured to me that if you are interested in a workaround for your particular problem, you can manually inset a comment after the first argument to force-wrap:
```
add_library(foo #
            src/file1.c
            src/file2.c)

```

**Comment by yeganer** [comment]
Thanks, that's a useable workaround. However it required adding the `#` at the end of every line, otherwise the files wouldn't wrap.

Like this:
```
add_library(foo #
                  src/file1.c #
                  src/file2.c)
```

**Comment by cheshirekow** [comment]
That's odd. I get wrapping with just the first line commented:
```
$ echo "add_library(foo #
> src/file1.c src/file2.c)" | cmake_format - 
add_library(foo #
            src/file1.c
            src/file2.c)
```
Can you share the configuration you're using?

**Comment by paulohefagundes** [comment]
I'm interested in this feature as well!

particularly in cases where we want to specify the targets whether they are PRIVATE, PUBLIC, or INTERFACE - I'd like for those to be in separate lines even if they're short enough for 1 line

examples:
```
target_sources(foo
    PRIVATE
       foo.cpp
    PUBLIC
       foo.hpp
)

target_link_libraries(foo
   PUBLIC
      bar
   PRIVATE
      baz
)
```

**Comment by slietzau** [comment]
> 
> 
> That's odd. I get wrapping with just the first line commented:
> 
> ```
> $ echo "add_library(foo #
> > src/file1.c src/file2.c)" | cmake_format - 
> add_library(foo #
>             src/file1.c
>             src/file2.c)
> ```
> 
> Can you share the configuration you're using?

I'm not sure which configuration required adding the second '#'. My current configuration produces the same behavior you observed.

**Comment by yeganer** [comment]
> 
> 
> That's odd. I get wrapping with just the first line commented:
> 
> ```
> $ echo "add_library(foo #
> > src/file1.c src/file2.c)" | cmake_format - 
> add_library(foo #
>             src/file1.c
>             src/file2.c)
> ```
> 
> Can you share the configuration you're using?

I'm not sure which configuration required adding the second '#'. My current configuration produces the same behavior you observed.

**Comment by cheshirekow** [comment]
between `max_pargs_hwrap` and forced-newlines I think the original ask here has been covered so I think I'm going to close this issue. However, we talked about a couple of ideas so I wanted to update you on progress:

*re: parg lists as specification*: There is now a lot more parsing logic and several cmake commands are parsed with custom logic. Several of these parsers output multiple positional argument lists (e.g. things like `[2, '*']` are implemented).  In the future I intend to allow custom parsers to be implemented in python. The "simple" legacy specification through a dictionary is still supported and I've got an issue on the internal issue tracker to implement "lists as parg specifications" which was your original proposed solution

*re: annotated file lists*: This is not currently implemented but a variant of it is. Positional argument groups in the parse tree can be "tagged" by the parser with a couple of different flags. Currently what is implemented is "sortable" and "default_wrapping". I've got an internal issue to make a decision about "is-file-list". I think that a configuration option like like "always wrap file lists" could be useful to some projects. That might be implemented by tagging a positional group as "is file list" or it might be implemented by assigning the "default_wrapping" based on the configuration setting at parse time. 

Feel free to continue discussion if you like and if you think there remains an open feature-request here we can re-open the issue.
